contrib/internal/httputil/trace.go

```diff
@@ -54,6 +58,34 @@ func TraceAndServe(h http.Handler, cfg *TraceConfig) {
 
 	cfg.ResponseWriter = wrapResponseWriter(cfg.ResponseWriter, span)
 
+	op := dyngo.StartOperation(
+		httpinstr.HandlerOperationArgs{
+			IsTLS:       cfg.Request.TLS != nil,
+			Method:      httpinstr.Method(cfg.Request.Method),
+			Host:        httpinstr.Host(cfg.Request.Host),
+			RequestURI:  httpinstr.RequestURI(cfg.Request.RequestURI),
+			RemoteAddr:  httpinstr.RemoteAddr(cfg.Request.RemoteAddr),
+			Headers:     httpinstr.Header(cfg.Request.Header),
+			UserAgent:   httpinstr.UserAgent(cfg.Request.UserAgent()),
+			QueryValues: httpinstr.QueryValues(cfg.Request.URL.Query()),
+		},
+	)
+	op.OnData(func(_ *dyngo.Operation, e *appsectypes.SecurityEvent) {
+		// Keep this trace due to the security event
+		span.SetTag(ext.SamplingPriority, ext.ManualKeep)
```

Comment:
dyngo's implementation removes all the operation listeners when it finishes. So it triggers before op.Finish() and span.Finish() deferred calls.

Reply:
Side note on this one: it actually triggers during the op.Finish() because this is when the WAF "flushes" the list of security events it monitored during the handler lifetime. This is made possible because an operation event listeners are removed after the operation finish event.